contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,594 @@
+        external
+        whenNotPaused
+        stakingPeriodComplete
+        rewardsAvailable
+        nonReentrant
+    {
+        User memory user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        delete _users[_msgSender()];
+
+        StakingPoolLib.Config storage _config = _stakingPoolConfig;
+
+        //slither-disable-next-line reentrancy-events
+        _transferStake(user.depositAmount, IERC20(_config.stakeToken));
+
+        _withdrawRewards(_config, user);
+    }
+
+    /**
+     * @notice Withdraw only stake tokens after staking period is complete. Reward tokens may not be available yet.
+     */
+    function withdrawStake()
+        external
+        stakingPeriodComplete
+        nonReentrant
+        whenNotPaused
+    {
+        User storage user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        uint128 currentDepositBalance = user.depositAmount;
+        user.depositAmount = 0;
+
+        StakingPoolLib.Config storage _config = _stakingPoolConfig;
+
+        // set users floating reward
+        for (uint256 i = 0; i < _config.rewardTokens.length; i++) {
+            if (_config.rewardType == StakingPoolLib.RewardType.FLOATING) {
+                user.rewardAmounts[i] = _calculateFloatingReward(
+                    _config.rewardTokens[i].ratio,
+                    currentDepositBalance
+                );
+            }
+        }
+
+        _transferStake(currentDepositBalance, IERC20(_config.stakeToken));
+    }
+
+    /**
+     * @notice Withdraw only reward tokens. Stake may have already been withdrawn.
+     */
+    function withdrawRewards()
+        external
+        stakingPeriodComplete
+        rewardsAvailable
+        whenNotPaused
+    {
+        StakingPoolLib.Config memory _config = _stakingPoolConfig;
+
+        User memory user = _users[_msgSender()];
+        require(user.depositAmount == 0, "StakingPool: stake first");
+        delete _users[_msgSender()];
+
+        bool noRewards = true;
+
+        for (uint256 i = 0; i < user.rewardAmounts.length; i++) {
+            if (user.rewardAmounts[i] > 0) {
+                noRewards = false;
+                //slither-disable-next-line calls-loop
+                _transferRewards(
+                    user.rewardAmounts[i],
+                    IERC20(_config.rewardTokens[i].tokens)
+                );
+            }
+        }
+        if (noRewards) {
+            emit NoRewards(_msgSender());
+        }
+    }
+
+    /**
+     * @notice Withdraw stake tokens when minimum pool conditions to begin are not met
+     */
+    function earlyWithdraw()
+        external
+        stakingPoolRequirementsUnmet
+        whenNotPaused
+    {
+        _withdrawWithoutRewards();
+    }
+
+    /**
+     * @notice Withdraw stake tokens when admin has enabled emergency mode
+     */
+    function emergencyWithdraw() external emergencyModeEnabled {
```

Comment:
This calls `_withdrawWithoutRewards` which will delete the caller's current reward data. This doesn't seem like what we want. Shouldn't we allow users to withdraw their deposit but leave their reward data in case the emergency is resolved?

Reply:
Agree, no reason to delete reward data unless the reward is claimed.

_If we keep the reward data, then it can be queried, allowing for manual rewarding if so desired, if we delete it then recalculation would be needed_

Reply:
Ok refactored to keep reward data